scope_registry.go

```diff
@@ -152,23 +152,51 @@ func (r *scopeRegistry) Subscope(parent *scope, prefix string, tags map[string]s
 	// as the memory layout of []byte is a superset of string the below casting is safe and does not do any alloc
 	// However it cannot be used outside of the stack; a heap allocation is needed if that string needs to be stored
 	// in the map as a key
-	if s, ok := r.lockedLookup(subscopeBucket, *(*string)(unsafe.Pointer(&buf))); ok {
-		subscopeBucket.mu.RUnlock()
-		return s
+	var (
+		unsanitizedKey = *(*string)(unsafe.Pointer(&buf))
+		sanitizedKey   string
+	)
+	if s, ok := r.lockedLookup(subscopeBucket, unsanitizedKey); ok {
+		// If this subscope isn't closed, or (in order to preserve historical
+		// behavior) if it isn't configured to report on close, return it.
+		if !s.closed.Load() || !s.reportOnClose {
+			subscopeBucket.mu.RUnlock()
+			return s
+		}
+
+		switch {
+		case parent.reporter != nil:
+			s.report(parent.reporter)
+		case parent.cachedReporter != nil:
+			s.cachedReport()
+		}
+
+		tags = parent.copyAndSanitizeMap(tags)
```

Comment:
Move these two lines out of the if/else block and get rid of else?

Reply:
Thanks - we still need to do some of this conditionally, but we can at least restructure this a bit so there's no duplication.